## Module
- A **module** is a single Python file (`.py`) that contains functions, classes, or variables you can reuse.
- Think of it as a **toolbox** with specific tools.
- Example:
  ```python
  import math
  print(math.sqrt(16))   # → 4.0
  ```
  Here, `math` is a module that provides mathematical functions.

---

##  Library
- A **library** is a collection of related modules packaged together.
- It’s like a **big toolkit** containing many smaller toolboxes (modules).
- Example:
  - **NumPy** → a library for numerical computing.
  - **Pandas** → a library for data analysis.
  ```python
  import numpy as np
  arr = np.array([1,2,3])
  print(arr.mean())   # → 2.0
  ```

---

##  Framework
- A **framework** is a much larger structure that provides not just tools, but also **rules, architecture, and workflows** for building applications.
- It’s like a **blueprint for building a house**—you follow its structure.
- Example:
  - **Django** → a web framework for building websites.
  - **Flask** → a lightweight web framework.
  ```python
  from flask import Flask
  app = Flask(__name__)

  @app.route("/")
  def home():
      return "Hello, World!"
  ```

---

##  Quick Analogy
- **Module** → A single screwdriver.
- **Library** → A full toolbox with screwdrivers, hammers, wrenches.
- **Framework** → A construction manual + toolbox + rules for building an entire house.

---

👉 In Python:
- **Modules** are the smallest unit (one file).
- **Libraries** are collections of modules.
- **Frameworks** are big systems built on top of libraries and modules, guiding how you build applications.

    `Module` → `Library` → `Framework`
  ---

# Math module

In [2]:
import math

In [3]:
print(math.sqrt(25))

5.0


In [5]:
print(math.pow(5,2))

25.0


In [12]:
from math import * # import all fuction without initial

In [13]:
print(pow(5,2))

25.0


In [14]:
pi

3.141592653589793

In [17]:
import math as m
print(m.nan)
print(m.floor(3.953))

nan
3


## Common Math Functions

| Function | Description | Example | Output |
|----------|-------------|---------|--------|
| `math.sqrt(x)` | Square root | `math.sqrt(16)` | `4.0` |
| `math.pow(x,y)` | Power (x^y) | `math.pow(2,3)` | `8.0` |
| `math.exp(x)` | Exponential (e^x) | `math.exp(2)` | `7.389...` |
| `math.log(x)` | Natural log (base e) | `math.log(10)` | `2.302...` |
| `math.log10(x)` | Log base 10 | `math.log10(100)` | `2.0` |
| `math.log2(x)` | Log base 2 | `math.log2(8)` | `3.0` |
| `math.ceil(x)` | Round up | `math.ceil(4.2)` | `5` |
| `math.floor(x)` | Round down | `math.floor(4.9)` | `4` |
| `math.trunc(x)` | Truncate decimals | `math.trunc(4.9)` | `4` |
| `math.factorial(x)` | Factorial | `math.factorial(5)` | `120` |
| `math.gcd(a,b)` | Greatest common divisor | `math.gcd(12,18)` | `6` |
| `math.lcm(a,b)` | Least common multiple | `math.lcm(12,18)` | `36` |
| `math.fabs(x)` | Absolute value | `math.fabs(-7.5)` | `7.5` |
| `math.isqrt(x)` | Integer square root | `math.isqrt(17)` | `4` |
| `math.comb(n,k)` | Combinations | `math.comb(5,2)` | `10` |
| `math.perm(n,k)` | Permutations | `math.perm(5,2)` | `20` |

---

## Constants in `math`
| Constant | Meaning | Example |
|----------|---------|---------|
| `math.pi` | π (3.14159…) | `math.pi` |
| `math.e` | Euler’s number (2.718…) | `math.e` |
| `math.inf` | Infinity | `math.inf` |
| `math.nan` | Not a number | `math.nan` |

---

## Trigonometric Functions
| Function | Description | Example | Output |
|----------|-------------|---------|--------|
| `math.sin(x)` | Sine (x in radians) | `math.sin(math.pi/2)` | `1.0` |
| `math.cos(x)` | Cosine | `math.cos(0)` | `1.0` |
| `math.tan(x)` | Tangent | `math.tan(math.pi/4)` | `1.0` |
| `math.asin(x)` | Inverse sine | `math.asin(1)` | `π/2` |
| `math.acos(x)` | Inverse cosine | `math.acos(0)` | `π/2` |
| `math.atan(x)` | Inverse tangent | `math.atan(1)` | `π/4` |
| `math.degrees(x)` | Convert radians → degrees | `math.degrees(math.pi)` | `180.0` |
| `math.radians(x)` | Convert degrees → radians | `math.radians(180)` | `3.14159...` |
